docs: add the design of table attributes (#28971)
pingcap/tidb: docs/design/2021-10-20-table-attributes.md

For the table, we can use the following SQL to add one or multiple attributes for the table `t`:

```sql
ALTER TABLE t ATTRIBUTES[=]'key=value[, key1=value1...]';
```

And reset it by using:

```sql
ALTER TABLE t ATTRIBUTES[=]DEFAULT;
```

For partition `p`, the way is almost the same as the table:

```sql
ALTER TABLE t PARTITION p ATTRIBUTES[=]'key=value[, key1=value1...]';
```

And reset it by using:

```sql
ALTER TABLE t PARTITION p ATTRIBUTES[=]DEFAULT;
```

The attributes here can be any string that doesn't contain the reserved words. 

There is no need for TiDB to know about the actual meaning of these attributes. The only thing that TiDB should do is to construct a request which contains the ID, the table/partition key range, the attributes, etc. Then synchronizing it with PD through the DDL job. All of these will be persisted in PD.

### Display Attributes

we add a new table in the `INFORMATION_SCHEMA` named `ATTRIBUTES`. If necessary, we can also support syntax like

```sql
SHOW ATTRIBUTES FOR [{DATABASE|SCHEMA} s][TABLE t][PARTITION p];
```
or 
```

> SHOW CREATE TABLE;

Reviewer: ```suggestion
SHOW CREATE TABLE t;
```

?

| schema/test/t1/p0 | key-range | "key=value"          | [7480000000000000ff3a5f720000000000fa, 7480000000000000ff3b5f720000000000fa]                                                                                                                                                             |
| schema/test/t1    | key-range | "key=value"          | [7480000000000000ff395f720000000000fa, 7480000000000000ff3a5f720000000000fa], [7480000000000000ff3a5f720000000000fa, 7480000000000000ff3b5f720000000000fa], [7480000000000000ff3b5f720000000000fa, 7480000000000000ff3c5f720000000000fa] |
+-------------------+-----------+----------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### DDL Management
Here are the concrete behaviors when doing the DDL operations with a table/partition having the attributes.

#### Drop Clause 

> + For the drop table operation, the attributes will be deleted once the GC has finished

Reviewer: ```suggestion
+ For the drop schema and table operations, the attributes will be deleted once the GC has finished
```

Author: Actually, when dropping schema, the attributes will be deleted immediately.

Reviewer: @rleungx well in that case update line 75 😓 

i think the attributes of the schema will be deleted immediately, but the attributes of the tables inside the schemas will need to wait for GC.

Author: updated

+ For the drop partition operation, the attributes will be deleted immediately

#### Truncate Clause
+ For the truncate table operation, both table range and partition range of attributes will be updated.
+ For the truncate partition operation, the partition range of attributes will be updated.

#### Recover/Flashback Clause

> + The attributes will be not changed once the recover/flashback process success.

Reviewer: ```suggestion
+ The attributes will be kept as the old values after the recover/flashback process succeed.
```

#### Exchange Clause
+ The attributes will be exchanged including the ID and ranges.

### Privilege Management

Privilege management is quite straightforward:

* `ALTER TABLE` statement requires `Alter` privilege
* `information_schema.attributes` only shows the attributes on the objects that visible to the current user
* `ADMIN SHOW DDL` requires `Super` privilege

### Inheritance

> Consider that if we add attributes for a table, all partitions of that table will inherit the attributes by default.

Reviewer: clarify the behavior of `CREATE TABLE x LIKE y`?

Author: For now, `x` won't have the same attributes as `y`.

Reviewer: yes i mean this is not written in the document. people shouldn't need to dig discussions around to know that.

Author: got
